### DISCUSSION TITLE
What objects are needed to construct an impact area scenario?

### DISCUSSION THREAD

**Discussion by rnugent3** [HydrologicEngineeringCenter/HEC-FDA#113] [Q&A] [answered]
Cody needs some help up front with making the needed revisions to support both Task 1 and incorporating a new model. To that end, we need to put a list together of the objects involved in constructing an impact area scenario. We then need to point to the corresponding object in the new model.

**Comment by rnugent3** [comment, marked as answer]
I am tracking the following objects:

1. Flow-frequency 
2. Inflow-outflow
3. Rating curve
4. Interior-exterior
5. Levee (including fragility function)
6. Stage-damage by damage category 
7. Damage-frequency 

An impact area scenario consists of the above relationships and the following properties:

1. Impact area ID
2. Analysis year
3. Threshold type
4. Threshold value
5. Name 
6. Description

**Reply by HenryGeorgist** [reply]
1. Flow Frequency is currently it's own class in fda-model. It contains an IDistribution (which will contain a sample size to describe the equivalent years of record), we have a path to making this code ready for testing, but we have not executed that work quite yet.
2. Inflow-Outflow is simply a PairedData or UncertainPairedData object
3. Rating Curve is simply a PairedData or UncertainPairedData object
4. InteriorExterior is simply a PairedData or UncertainPairedData object
5. Levee is simply a PairedData or UncertainPairedData object
6. Stage-Damage is simply a PairedData or UncertainPairedData object
7.  Damage Frequency is computed internally to the ead.simulation.compute method right now, it is a PairedData object (because it is what we integrate)

We have this as our impact area class: https://github.com/HydrologicEngineeringCenter/fda-model/blob/main/fda-model/impactarea/ImpactArea.cs
It contains a name and a simulation.

We have this as our Scenario class: https://github.com/HydrologicEngineeringCenter/fda-model/blob/main/fda-model/scenarios/Scenario.cs

It contains a year and a list of impact areas.

A simulation itself (contained by an impact area) contains the results object which would track AEP (and the AEP threshold) and EAD.

It seems we need to add Name and Description (probably to scenario?) and Impact AreaID to impact area.

We might need to add Threshold Type for non stage related thresholds.

**Reply by rnugent3** [reply]
Is the object created by the scenario class an impact area scenario? That object should represent one impact area, not all impact areas. 

I appreciate that this design is not good, it reflects the design of a "condition" in the original codebase

**Reply by rnugent3** [reply]
For concreteness, Cody has the following objects that are required for an impact area scenario 
`            ConditionsElement elem = new ConditionsElement(name, description, year, impAreaID, flowFreqID, infOutflowID,
                ratingID, extIntID, leveeFailureID, stageDamageID, thresholdType, thresholdValue);`

**Reply by rnugent3** [reply]
Hi @HenryGeorgist how is progress going on tightening up the scenario object? @codymccoy could you tell us a little more information about your requirements to proceed on this part of the task?

**Reply by HenryGeorgist** [reply]
@rnugent3 I have no significant progress to report. I need some clarity on the direction here - should I adapt my code to have one impact area id per condition or should I keep my current list of impact areas per condition?

**Reply by HenryGeorgist** [reply]
i have added stage damage by damage category to the simulation (and results) objects.  
https://github.com/HydrologicEngineeringCenter/fda-model/blob/main/fda-model/ead/simulation.cs
https://github.com/HydrologicEngineeringCenter/fda-model/blob/main/fda-model/metrics/Results.cs

**Comment by rnugent3** [comment]
The threshold types currently are interior stage, exterior stage, and damages.

**Reply by HenryGeorgist** [reply]
That is cool, that would go into: https://github.com/HydrologicEngineeringCenter/fda-model/blob/main/fda-model/metrics/IContainResults.cs

in that class we would need to specify if AEP is damage, interior stage or exterior stage.

**Reply by HenryGeorgist** [reply]
@rnugent3 I recommend we do not add exterior stage, just interior stage and damage

**Reply by HenryGeorgist** [reply]
whatever is consistent with our MVP FDA 1.2.4 functionality and no features unless coerced...

**Reply by rnugent3** [reply]
I think all three are relevant. In a meeting with Cody yesterday, I noticed that the threshold types are listed as damages, EAD, interior stage, and exterior stage. I was not tracking all four. I think damages are the current default threshold in FDA 1. EAD and interior stage seem to be the thresholds new to FDA (I am not positive how CNEP and AEP are calculated in the case of an interior-exterior relationship in FDA 1). We will already be collecting the EAD so an EAD threshold is easy.

**Reply by HenryGeorgist** [reply]
got it. The levee component makes perfect sense.

**Comment by rnugent3** [comment]
It has to be one impact area id per condition because we have one flow frequency per condition, one rating curve, so on and so forth.

**Comment by rnugent3** [comment]
I am trying to think of a use case here. If we want to know the assurance of a levee, don’t we want to use exterior stage? If we have an exterior-interior relationship, then exterior != interior with overtopping, but overtopping at the exterior stage is considered a mode of failure. What do you think?
